### PR TITLE
Jseldess/data types

### LINKED ISSUE #17: data types
Should be part of the SQL Reference section.

### PR DESCRIPTION
**Pull request by jseldess**
First draft of page on CockroachDB data types. @nvanbenschoten, @petermattis, @spencerkimball, @bdarnell, please let me know what you think. I don't think the explanation and examples for `BYTES` are sufficient, so please send ideas there.

HTML version here: http://cockroach-draft-docs.s3-website-us-east-1.amazonaws.com/data-types.html

Fixes #17 and #123

### CODE REVIEW

File: data-types.md [hunk 1]
```diff
@@ -3,4 +3,408 @@ title: Data Types
...
+[`BOOL`](#bool) | A Boolean value. | `true` 
+[`DATE`](#date) | Year, month, day. | `DATE '2016-01-25'`
+[`TIMESTAMP`](#timestamp) | A date and time pairing. | `TIMESTAMP '2016-01-25 10:10:10'`
+[`INTERVAL`](#interval) | A span of time. | `INTERVAL '2h30m30s'`
+[`STRING`](#string) | A string of characters. | `'a1b2c3'`
+[`BYTES`](#bytes) | A string of binary characters. | `'a1b2c3'`
+
+## `INT`
+
+The `INT` type stores 64-bit signed integers, that is, whole numbers from 9,223,372,036,854,775,808 to 9,223,372,036,854,775,807.
```
Comment (bdarnell) [inline data-types.md]: The first of these needs a minus sign.

File: data-types.md [hunk 2]
```diff
@@ -3,4 +3,408 @@ title: Data Types
...
+
+#### Synonyms 
+
+In CockroachDB, the following are synonyms of `INT` and are implemented identically: 
+
+- `SMALLINT` 
+- `INTEGER` 
+- `INT64` 
+- `BIGINT`
+- `BITS`
```
Comment (bdarnell) [inline data-types.md]: I don't think `BITS` exists.

File: data-types.md [hunk 3]
```diff
@@ -3,4 +3,408 @@ title: Data Types
...
+Precision (the maximum count of digits in a whole number, both to the left and right of the decimal point) and scale (the maximum count of digits to the right of the decimal point) are always unlimited for a `DECIMAL` value. If you specify these limits by declaring `DECIMAL(precision, scale)` on a column, they will not be enforced. 
+
+#### Format
+
+When declaring a decimal, format it as `DECIMAL '1.2345'`. This casts the value as a string, which preserves the number's exact precision.
+
+Alternately, you can cast a `FLOAT` as a `DECIMAL`, but note that precision will be limited to 17 digits in total (both to the left and right of the decimal point):
+
+- `1.2345::DECIMAL`
+- `CAST (1.2345 AS DECIMAL)`
```
Comment (bdarnell) [inline data-types.md]: No spaces between `CAST` and the parentheses. I would only use the `CAST` syntax in these examples, and in one place (probably in the documentation for the `CAST` function) say that `X::Y` is synonymous with `CAST(X AS Y)`.

File: data-types.md [hunk 4]
```diff
@@ -3,4 +3,408 @@ title: Data Types
...
+
+The `BOOL` type stores a Boolean value of `false` or `true`. 
+
+#### Synonyms
+
+In CockroachDB, `BOOLEAN` is a synonym of `BOOL` and is implemented identically. 
+
+#### Format
+
+When declaring a `BOOL`, format it as `false` or `true`.
```
Comment (bdarnell) [inline data-types.md]: Mention that `true` and `false` are case-insensitive.

File: data-types.md [hunk 5]
```diff
@@ -3,4 +3,408 @@ title: Data Types
...
+#### Format
+
+When declaring a `DATE`, format it as `DATE '2016-01-25'`. 
+
+Alternately, you can format a date as follows:
+
+- `'2016-01-25'::DATE`
+- `CAST ('2016-01-25' AS DATE)` 
+
+Note that a date is stored with hour, minute, second, and timezone set to 0.
```
Comment (bdarnell) [inline data-types.md]: Dates don't really have these fields, so it's confusing to talk about them being stored that way. Rather than talk about how things are stored, say something like "in some contexts, dates may be displayed with hours, minutes, seconds and timezone set to zero". (it's kind of a bug whenever this happens, but it's harder to fix than it looks)

File: data-types.md [hunk 6]
```diff
@@ -3,4 +3,408 @@ title: Data Types
...
+## `TIMESTAMP`
+
+The `TIMESTAMP` type stores a date and time pair in UTC. 
+
+#### Format
+
+When declaring a `TIMESTAMP`, use one of the following formats:
+
+- Date only: `TIMESTAMP '2016-01-25'`
+- Date and Time: `TIMESTAMP '2016-01-25 10:10:10.999999999'`
```
Comment (bdarnell) [inline data-types.md]: The fractional seconds part is optional in each of these cases.

File: data-types.md [hunk 7]
```diff
@@ -3,4 +3,408 @@ title: Data Types
...
+#### Synonyms
+
+In CockroachDB, the following are synonyms of `BYTES` and are implemented identically: 
+
+- `BYTEA` 
+- `BLOB` 
+
+#### Format
+
+When declaring a `BYTES` string, format it as `'a1b2c3'`.
```
Comment (bdarnell) [inline data-types.md]: That's still a `STRING`, not a `BYTES`. Sometimes you can use them interchangeably, but not always. A better way to make `BYTES` is to put a `b` in front of the quotes: `b'a1b2c3'`

File: data-types.md (hunk 1 above)
Comment (jseldess) [inline data-types.md]: Thanks, Ben. Fixed with next commit.

File: data-types.md (hunk 2 above)
Comment (jseldess) [inline data-types.md]: Ah, this should be `BIT` instead. I see this in [types_test.go](https://github.com/cockroachdb/cockroach/blob/master/sql/parser/types_test.go#L30). And when I set a column to `bit`, it accepts that and converts it to `int`, just like the other synonyms:

```
root@:26257> create table bits (a bit, b int);
CREATE TABLE
root@:26257> show columns from bits;
+-------+------+-------+----------------+
| Field | Type | Null  |    Default     |
+-------+------+-------+----------------+
| a     | INT  | true  | NULL           |
| b     | INT  | true  | NULL           |
| rowid | INT  | false | unique_rowid() |
+-------+------+-------+----------------+
```

OK to leave this in as `BIT`?

File: data-types.md (hunk 3 above)
Comment (jseldess) [inline data-types.md]: OK, I'lll remove the `X::Y` syntax and mention it when I get to the `CAST` function. Fixed with next commit.

File: data-types.md (hunk 4 above)
Comment (jseldess) [inline data-types.md]: Fixed with next commit.

File: data-types.md (hunk 5 above)
Comment (jseldess) [inline data-types.md]: Fixed with next commit.

File: data-types.md (hunk 6 above)
Comment (jseldess) [inline data-types.md]: @bdarnell, aren't both seconds and fractional seconds optional?

File: data-types.md (hunk 7 above)
Comment (jseldess) [inline data-types.md]: Fixed with next commit.

File: data-types.md [hunk 8]
```diff
@@ -3,4 +3,391 @@ title: Data Types
...
+-----|-------------|--------
+[`INT`](#int) | A 64-bit signed integer. | `12345`
+[`DECIMAL`](#decimal) | An exact, fixed-point number. | `DECIMAL '1.2345'`
+[`FLOAT`](#float) | An inexact, floating-point number. | `1.2345`
+[`BOOL`](#bool) | A Boolean value. | `true` 
+[`DATE`](#date) | Year, month, day. | `DATE '2016-01-25'`
+[`TIMESTAMP`](#timestamp) | A date and time pairing. | `TIMESTAMP '2016-01-25 10:10:10'`
+[`INTERVAL`](#interval) | A span of time. | `INTERVAL '2h30m30s'`
+[`STRING`](#string) | A string of characters. | `'a1b2c3'`
+[`BYTES`](#bytes) | A string of binary characters. | `'a1b2c3'`
```
Comment (nvanbenschoten) [inline data-types.md]: Same comment as Ben below. Bytes needs to be escaped, as discussed yesterday with the `b'a1b2c3'` syntax. 

I would also refer to this below along the lines of "byte escaping" or "byte escape format". It has a different connotation than plain "formatting".

**Comment by nvanbenschoten** [comment]
Two more general comments that we may want to break off into separate issues.
- Clicking on section title links still pulls the title partially under the header
- While skimming over the page quickly, it's pretty difficult to determine the hierarchy of information because the title and subtitles all look so similar (h2 vs. h4 text, but the same color). I would prefer these to be a little more easily distinguishable from each other.

File: data-types.md (hunk 2 above)
Comment (bdarnell) [inline data-types.md]: Hmm, I'd have expected `BIT` to be synonymous with `BOOL`, not `INT`. Postgres doesn't have a `BIT` type; mysql has one but it's parameterized (`BIT(n)`, which makes it more clear that it's a numeric type instead of a single bit). It looks like we support that too although we don't do anything with the parameter. We should document `BIT(n)` with the parameter instead of just saying that it's equivalent to `INT`.

**Comment by nvanbenschoten** [comment]
I'd be curious about other people's input, but I find it a little strange to find scattered mentions of available casts on the "data types" page. As a developer, I'd rather have the page focus on the strictly idiomatic way to use each type, and have all supported casts on a different page which I can look at whenever I need to wrangle with types.

File: data-types.md (hunk 6 above)
Comment (bdarnell) [inline data-types.md]: No; if you give hours and minutes you have to give seconds too.

File: data-types.md (hunk 8 above)
Comment (jseldess) [inline data-types.md]: Sorry I somehow missed that in our talk yesterday, Nate. Fixing with next commit.

**Comment by jseldess** [comment]
@nvanbenschoten, jumping to an anchor link has been behaving like that for a while. I agree that it's annoying. I just opened docs issue #133. 

I also agree that this single, long page is hard to parse. I'll split each data type onto its own page, as we did with sql statements.

File: data-types.md (hunk 6 above)
Comment (jseldess) [inline data-types.md]: OK, fixed with the next commit.

**Comment by nvanbenschoten** [comment]
SGTM, thanks @jseldess. Everything else seems to be looking good.

**Comment by jseldess** [comment]
OK, each type now on its own page, and I removed most mentions of cast, excepting cases where there's something important to call out (cast infinity and NaN for float, cast 0 or 1 for boolean, cast float as decimal limits precision).

**Comment by nvanbenschoten** [comment]
LGTM

File: data-types.md (hunk 2 above)
Comment (jseldess) [inline data-types.md]: Sorry I didn't get this into this PR. I'll open a new one. But I'm a little confused. It looks to me like `BIT(n)` is really equivalent to `INT`. You can declare a column as `BIT(5)` when creating a table, and it'll show up as `INT(5)` on `show columns from`, but we don't enforce the limit. 

```
root@:26257> create table bits (a bit(5), b bit, c int);
CREATE TABLE
root@:26257> show columns from bits;
+-------+--------+-------+----------------+
| Field |  Type  | Null  |    Default     |
+-------+--------+-------+----------------+
| a     | INT(5) | true  | NULL           |
| b     | INT    | true  | NULL           |
| c     | INT    | true  | NULL           |
| rowid | INT    | false | unique_rowid() |
+-------+--------+-------+----------------+
root@:26257> insert into bits values (123456, 1, 2);
INSERT 1
root@:26257> select * from bits;
+--------+---+---+
|   a    | b | c |
+--------+---+---+
| 123456 | 1 | 2 |
+--------+---+---+
```

Maybe I should list `BIT(n)` as a synonym for `INT` and also have a section on length and state that the limit won't be enforced, similar to what we say for [`STRING`](https://www.cockroachlabs.com/docs/string.html)?

Comment (bdarnell) [inline data-types.md]: `BIT` may be synonymous with `INT` in our current implementation, but then that's a bug. `INT(5)` is supposed to be a 5-digit number (up to 99999), while `BIT(5)` is a 5-bit number (up to 31). I would just leave it undocumented until we can give it a better implementation.

Comment (jseldess) [inline data-types.md]: OK, I'll remove `BIT` then. Just open an issue for this: https://github.com/cockroachdb/cockroach/issues/5514

File: bytes.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
+---
+title: BYTES
+toc: true
+---
+
+## Description
+
+The `BYTES` [data type](data-types.html) stores binary strings of variable, unlimited length.
```
Comment (petermattis) [inline bytes.md]: While the length isn't currently limited, it is bold to claim `unlimited length`. It is likely a bad idea to store a value that is larger than the range size (64 MB). I'm not exactly sure how to phrase the length recommendation.

File: decimal.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+toc: true
+---
+
+## Description
+
+The `DECIMAL` [data type](data-types.html) stores exact, fixed-point numbers with no limit on digits to the left and right of the decimal point. This type is used when it is important to preserve exact precision, for example, with monetary data. 
+
+## Synonyms
+
+In CockroachDB, the following are synonyms of `DECIMAL` and are implemented identically:
```
Comment (petermattis) [inline decimal.md]: The `and are implemented identically` phrase used here and elsewhere seems to be implied by `synonym`. Do you think the additional detail is useful?

File: decimal.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+## Precision and Scale
+
+Precision (the maximum count of digits in a whole number, both to the left and right of the decimal point) and scale (the maximum count of digits to the right of the decimal point) are always unlimited for a `DECIMAL` value. If you specify these limits by declaring `DECIMAL(precision, scale)` on a column, they will not be enforced. 
+
+## Format
+
+When declaring a decimal, format it as `DECIMAL '1.2345'`. This casts the value as a string, which preserves the number's exact precision.
+
+Alternately, you can cast a float as a decimal: `CAST(1.2345 AS DECIMAL)`. However, note that precision will be limited to 17 digits in total (both to the left and right of the decimal point).
```
Comment (petermattis) [inline decimal.md]: Should we note in the documentation that needing to cast a string to a decimal and the loss of precision when casting from a float are bugs?

File: string.md [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+- `CHAR` 
+- `CHAR(n)` 
+- `VARCHAR`
+- `VARCHAR(n)` 
+- `TEXT`
+
+## Length
+
+Length is always variable and unlimited for a `STRING` value. If you specify a fixed length by declaring `CHAR(n)` or `VARCHAR(n)` on a column, it will not be enforced.
```
Comment (petermattis) [inline string.md]: The length is supposed to be supported. That we're not handling this correctly for columns is a bug. Consider:

```
root@:26257> select 'hello world'::char(5);
+--------------------------------+
| CAST('hello world' AS CHAR(5)) |
+--------------------------------+
| hello                          |
+--------------------------------+
```

File: bytes.md (hunk 9 above)
Comment (JackKrupansky) [inline bytes.md]: And if 64MB (or is that MiB? And it probably needs to be reduced by the overhead of the key) becomes the hard limit, I would suggest that it should probably be recommended to use less than half of that (32MB) to allow ranges to be split into at least two non-empty ranges.

And there may be practical reasons to keep individual row/column values even smaller, like below 1 MB. Large BLOB support is a distinct beast from normal transactional records.

Comment (jseldess) [inline bytes.md]: Good points. For now, I'll just state "variable length". I opened an issue to figure out the exact recommendations (probably post-beta release): issue #159

File: decimal.md (hunk 10 above)
Comment (jseldess) [inline decimal.md]: Nope. I'll remove that wherever I've used it.

File: decimal.md (hunk 11 above)
Comment (jseldess) [inline decimal.md]: Adding a note that we'll supporting literals in an upcoming version.

File: string.md (hunk 12 above)
Comment (jseldess) [inline string.md]: If it's ok, I'll just remove "unlimited" and keep as-is for now. Once we've fixed this, I'll updated.
